Bonus guide: Balance of Satoshis

VajraOfIndra opened the pull request:
#### What

This PR is a new lightning bonus guide for Balance of Satoshis (BoS), rewritten for and tested on v3.

The guide explains:

1. how to install, upgrade and uninstall BoS
2. how to use it to rebalance channels
3. how to set up a LN node monitoring Telegram bot connected to the node

#### Why

BoS is a widely used tool by LN node operators. ThunderHub uses BoS in the background for its rebalancing tool. It is very useful for rebalancing channels (its original purpose) but contains many additional utilities such as a way to create a Telegram bot that helps monitoring the node and provides redundant storage for the SCB backup files.

#### How

Create a new file in /bonus/lightning
Updated /bonus/lightning.index.md to add the bonus guide under a new header 'Liquidity management (rebalancing)

#### Scope

- [ ] significant change to core configuration
- [x] independent bonus guide
- [ ] simple bug fix

#### Test & maintenance

Test: Follow the guides after having installed LND
Maintenance: 

- occasionally checking that the external links are not dead
- Updating the quote NodeJS version if updated in main guide for the Explorer

#### Troubleshooting / keep an eye on...

1) For the NodeJS update section, I'm not entirely sure that this is the best way to do this. We could also simply delete this section and assume that users will have a version younger than 12 or else none at all and use the install from scratch section...

2) I get an error message when torifying the` bos telegram `command. The command still works but I wonder if the communication to the Telegram server is done via Tor or not.

```
$ /usr/bin/torify bos telegram
> 1638539250 WARNING torsocks[4613]: [syscall] Unsupported syscall number 291. Denying the call (in tsocks_syscall() at syscall.c:604)
```
Does someone know how to fix this or does it need to be fixed?

#### Illustration

Telegram bot output example with rebalancing events from using `bos rebalance`:
![TGbot](https://github.com/VajraOfIndra/RaspiBolt/blob/pics/images/tgbot.png?raw=true)

![balance](https://media.giphy.com/media/NOnEwj4xBaPbG/giphy.gif)

Stadicus commented on bonus/lightning/balance-of-satoshis.md:
````diff
@@ -0,0 +1,369 @@
+
+  ```sh
+  $ sudo su - bos
+  $ npm i -g balanceofsatoshis
````
Wouldn't you need to update the GitHub repository first?

VajraOfIndra replied:
Good point... Rather than updating, Alex recommends to reinstall BoS:
![image](https://user-images.githubusercontent.com/86242283/144706081-0c5cbf25-3a24-4d5a-8f67-f6a6e6eedf68.png)

Does it mean we should just do this:

```
$ sudo su - bos
$ git clone https://github.com/alexbosworth/balanceofsatoshis.git /home/bos/balanceofsatoshis
$ cd balanceofsatoshis
$ npm install -g balanceofsatoshis
$ cd ~/
```
Is it better to delete the /home/bos/balanceofsatoshis folder first?

Wdyt is the best way to handle this?

Stadicus replied:
I don't think there's a difference between deleting the whole repo & cloning it again, or just pulling the lastest release. Unless I miss something with NPM...

My guess is that following the same steps as used with RTL for example should be the most efficient way.

```bash
bos@raspibolt3:~/balanceofsatoshis $ git fetch
bos@raspibolt3:~/balanceofsatoshis $ git describe --tags --abbrev=0
v11.13.0
bos@raspibolt3:~/balanceofsatoshis $ git checkout v11.13.0
Previous HEAD position was 07f6db4 fix typo
HEAD is now at 1b2a38d add docs for coop close on open
bos@raspibolt3:~/balanceofsatoshis $ git verify-tag v11.13.0
gpg: Signature made Thu 02 Dec 2021 19:55:14 GMT
gpg:                using RSA key DE23E73BFA8A0AD5587D2FCDE80D2F3F311FD87E
gpg: Good signature from "Alex Bosworth <alex.bosworth@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: DE23 E73B FA8A 0AD5 587D  2FCD E80D 2F3F 311F D87E
bos@raspibolt3:~/balanceofsatoshis $ npm install
```

Notes:
* Here, the command `git describe --tags --abbrev=0` is necessary. It's more stable, I need to upgrade the other guides as well.
* Importing Alex's PGP keys are required befor the `verify-tag` works:
  ```sh
  $ gpg --receive-keys E80D2F3F311FD87E
  ```

Stadicus replied:
We might want to add the `git reset --hard HEAD` command before checking out the new version, just to be sure. See https://github.com/raspibolt/raspibolt/issues/826#issuecomment-989164753

VajraOfIndra replied:
I updated the 'Upgrade' section using the RTL example + the git reset command.

I added the gpg --receive-keys command to the initial installation section as it has to be done once only.

I'm trying to make sense of the gpg --receive keys command. When I check Alex's keybase.io profile, I see the main GPG key ID on the left (although there is also a SF at the end of it that I don't know what it is..). Why can't we use the proposed command in his profile: "curl https://keybase.io/alexbosworth/pgp_keys.asc | gpg --import"? I tried it and it doesn't verify the tag, it seems to be a different key altogether... could you help me clarify this?

Stadicus commented on bonus/lightning/balance-of-satoshis.md:
````diff
@@ -0,0 +1,369 @@
+  $ userdel -r bos
+  ``` 
+  
+## Optional: connect your node to a Telegram bot
````
In general, I add a horizontal line before H2 headings with 

```
---
```

VajraOfIndra replied:
Fixed. I'll try to keep this in mind for the future guides.

Stadicus reviewed (changes requested):
First part of the review, including installation to running `bos peers`.

In general, what's the reasoning to install it with `npm global`? I installed it without it (like the BTC RPC Explorer), and it worked equally well (the binary file is then `/home/bos/balanceofsatoshis/bos`).

In general, I'm not very experienced in rebalancing. Maybe another review from someone more familiar with node operations would be helpful?

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
```diff
@@ -0,0 +1,369 @@
+{: .no_toc }
+
+Difficulty: Easy
+{: .label .label-green }
```
> I think it would be great to add a "Status: Tested V3" label to new (and updated) bonus guides as well.

Done!

> When I wrote the Boot from microSD bonus guide, I realized that multiple labels next to each other are only neatly aligned if there's some text before. To get some structure, I finally settled on this layout:

That's good to know, I moved the intro text before the labels.. I might shorten the intro if too long.

Stadicus replied:
Ack

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
```diff
@@ -0,0 +1,369 @@
+Difficulty: Easy
+{: .label .label-green }
+
+<details open markdown="block">
```
> I'm not convinced that the collapsible ToC adds value, it just complicates the markdown file in my opinion.

Ok, I removed it.

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
```diff
@@ -0,0 +1,369 @@
+
+## Installation
+
+### Check NodeJS
```
There is already a link to RTL for users that don't have NodeJS installed on their node. 

I removed the update NodeJS section to simplify the guide as users with a NodeJS version older than v12 shouldbe pretty rare so it's not that needed.

But not sure I got what you meant exactly, could you clarify how you'd like this section handled?

Stadicus replied:
Looks good :)

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
````diff
@@ -0,0 +1,369 @@
+
+  ```sh
+  $ sudo adduser --disabled-password --gecos "" bos
+  $ sudo /usr/sbin/usermod --append --groups lnd bos
````
Doesn't the command also just add the user "bos" to only one group, i.e. "lnd"?

But yeah the adduser command is much simpler in syntax, I changed accordingly.

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
````diff
@@ -0,0 +1,369 @@
+* NojeJS v1.0 or above should have been installed for the BTC RPC Explorer. We can check our version of NodeJS with user "admin": 
+
+  ```sh
+  $ node-v
````
Good catch!

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
```diff
@@ -0,0 +1,369 @@
+
+A good illustration is provided in Chapter 5 of 'Mastering the Lighnting Network' by Antonopoulos *et al.* ([source](https://github.com/lnbook/lnbook/blob/ec806916edd6f4d1b2f9da2fef08684f80acb671/05_node_operations.asciidoc#rebalancing-channels){:target="_blank"}) (Alice (A), is your node, Bob (B) and and Chan (C) are two peers with whom you have an opened channel with):
+
+![circular-rebalancing](https://github.com/lnbook/lnbook/raw/ec806916edd6f4d1b2f9da2fef08684f80acb671/images/mtln_0504.png)
```
I don't have Photoshop, can I do this with Gimp for future guides..?

Fixed (I use this pic and placed it in  the /images folder).

Stadicus replied:
I used GIMP for that myself (don't want to always start the Windows VM... :laughing: ). I can't add the file here, will send it to you so that you can add it under `/images/sources`.

VajraOfIndra replied:
Thanks for the file. I updated the file, it should appear with transparent background now. PTAL.

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
```diff
@@ -0,0 +1,369 @@
+  $ mkdir /home/bos/.npm-global
+  $ npm config set prefix '/home/bos/.npm-global'
+  $ exit
+  $ sudo bash -c "echo 'PATH=$PATH:/home/bos/.npm-global/bin' >> /home/bos/.bashrc"
```
Yeah, when I went again through the guide i wondered why I did that! But left it just in case I forgot there was a good reason.

Fixed now!

VajraOfIndra replied:
Fixed, added the source command below the echo export command.

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
````diff
@@ -0,0 +1,369 @@
+* Download the source code and install it
+  
+  ```sh
+  $ sudo su - bos
````
Nice, I changed accordingly.

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
````diff
@@ -0,0 +1,369 @@
+  
+  ```sh
+  $ sudo su - bos
+  $ git clone https://github.com/alexbosworth/balanceofsatoshis.git /home/bos/balanceofsatoshis
````
Ok, fixed

Stadicus commented on bonus/lightning/balance-of-satoshis.md:
```diff
@@ -0,0 +1,369 @@
+  $ mkdir /home/bos/.npm-global
+  $ npm config set prefix '/home/bos/.npm-global'
+  $ exit
+  $ sudo bash -c "echo 'PATH=$PATH:/home/bos/.npm-global/bin' >> /home/bos/.bashrc"
```
This still needs to be addressed

VajraOfIndra replied:
> See more on the [`export` command](https://linuxhandbook.com/export-command/).

Thanks for the link. I took some time to understand better what is happening there, I think I get it now, woohoo. I updated the command accordingly and added the source command. The .npm-global thing will be addressed in another comment.

VajraOfIndra commented:
> First part of the review, including installation to running `bos peers`.

Thanks!

> In general, what's the reasoning to install it with `npm global`? I installed it without it (like the BTC RPC Explorer), and it worked equally well (the binary file is then `/home/bos/balanceofsatoshis/bos`).

I updated the guide accordingly and tested it on my node. Seems to work well as well. I'm assuiming the global install was for bos commands to be used by other users such as admin? I asked on the Telegram group but no reply yet.
 
> In general, I'm not very experienced in rebalancing. Maybe another review from someone more familiar with node operations would be helpful?

That could be great. Since the other reviewers seem to be a busy right now, we could maybe merge the guide as is and ask for a review of the rebalancing section later on..? wdyt?

For the update section, it didn't work on my previous setup with the global install. I now run the latest version locally so can't test the upgrade section until there is a new version available. I'll keep an eye on that to test the upgrade section again.
I tried to describe what each git command does.. PTAL and correct any misunderstandings or poor wording.

I also added a section at the end that explains how the Bos bot work, PTAL!

The whole guide is quite long and has some advanced features (rebalancing using CLI), I wonder if we should put the 'Intermediate' difficulty tag for this guide? (tbh, I'm not entirely sure what qualify as 'Simple' vs 'Intermediate' vs 'Difficult'.. is it the setup process? or the use of the program? or something else?)

Thanks!

VajraOfIndra commented on bonus/lightning/balance-of-satoshis.md:
````diff
@@ -0,0 +1,369 @@
+  $ ls -la
+  ```
+
+* Set up nmp-global
````
See reply to your main comment. 

The npm-global set up is recommended in all the guides I could found (Alex's node guide, Raspiblitz, Plebnet wiki). I'm nit sure why, but I assume it is for other users to be able to use the bos commands..? I asked on the BoS Telegram group but no reply yet. Will keep you posted if I get some useful rationales.

I removed it for now as you suggested and install BoS locally with the bos user.

Stadicus commented on bonus/lightning/balance-of-satoshis.md:
```diff
@@ -0,0 +1,500 @@
+
+[Balance of Satoshis](https://github.com/alexbosworth/balanceofsatoshis){:target="_blank"} (BoS), created and maintained by LND developper Alex Bosworth, is a tool to work with LND. The most used feature of BoS is the rebalancing command that allows to manage your channels liquidity (which is the main focus of this guide), but BoS contains now many other lightning tools outside liquidity manageme ...
+
+Difficulty: Easy
```
Classifying this as "intermediate" makes definitely sense

VajraOfIndra replied:
Done, updated to medium/yellow label

Stadicus commented on bonus/lightning/balance-of-satoshis.md:
````diff
@@ -0,0 +1,500 @@
+  > [...]
+  ```  
+
+* Check the version with the --version (or -V) option
````
you could put the arguments in backticks, like "with the `--version` option". Also I think one argument is enough, makes it more readable

VajraOfIndra replied:
Yes, I tend to go on and on and on, I need to learn to simplify my writing :smile: 
Fixed, I added the backticks and removed --version to just keep -V

Stadicus commented on bonus/lightning/balance-of-satoshis.md:
```diff
@@ -0,0 +1,500 @@
+
+* Install Balance of Satoshis locally
+
+  $ npm install
```
Backticks are missing for this code block

VajraOfIndra replied:
Well spotted! fixed!

Stadicus commented on bonus/lightning/balance-of-satoshis.md:
````diff
@@ -0,0 +1,500 @@
+ 
+  ```sh
+  $ git fetch
+  > remote: Enumerating objects: 29, done.
````
Are you sure the full output is necessary? `git fetch` just updates the metadata and should never run into issues.

VajraOfIndra replied:
Good point, I reduced it to display only the last line with the new tag

Stadicus reviewed (changes requested):
It seems that all titles are one level too low:

* Bonus guide: Balance of Satoshis --> should be H1 (`#`)
* Requirements --> should be H2 (` ##`)
* etc...

Other than that, just a few little things, then I think its ready for merging

VajraOfIndra commented:
> It seems that all titles are one level too low:
> 
>     * Bonus guide: Balance of Satoshis --> should be H1 (`#`)
> 
>     * Requirements --> should be H2 (` ##`)
> 
>     * etc...

Quite a few guides are like this, I thought it was on purpose. Good to know, I'll update my other PRs accordingly.

> Other than that, just a few little things, then I think its ready for merging

All fixed now I believe, see comments below and PTAL (I also added a top illustration and a link to the BOS Telegram group for support)

Stadicus commented:
> Quite a few guides are like this, I thought it was on purpose. Good to know, I'll update my other PRs accordingly.

Oh, I just took the main guide for reference, because I felt the headers became really tiny... :)

Stadicus reviewed (approved):
LGTM! :tada:

Geochronist commented:
> Oh, I just took the main guide for reference, because I felt the headers became really tiny... :)

Agreed, and like this it's more consistent across all the guides.

VajraOfIndra commented:
> > Quite a few guides are like this, I thought it was on purpose. Good to know, I'll update my other PRs accordingly.
> 
> Oh, I just took the main guide for reference, because I felt the headers became really tiny... :)

You're right and like this it's more consistent across all guides.
